### PR TITLE
Confidence ensembles implementation

### PR DESCRIPTION
**Pull request by Kipok**
# What does this PR do ?

Adds confidence ensembles implementation.

I will follow up with more features (e.g., tuning of the LR/confidence parameters) as well as documentation and probably a tutorial notebook. But the basic functionality is fully implemented.

**Collection**: ASR

# Changelog 
- Add new script to `scripts/confidence-ensembles/build_ensemble.py`. It can be run to build an ensemble by providing models/data. It will train confidence-based model selection block and output .nemo checkpoint containing all models + pkl file for the model selection.
- Add new argument to `transcribe_speech.py` `return_transcriptions` which can be used to select if transcriptions or config is returned. Default to current behavior of returning config.
- Add a new temperature parameter to the decoding strategies and ctc/rntt decoder/joint blocks. What's currently called `temperature` in the confidence config is not really a temperature, but instead an `alpha` parameter as defined here https://arxiv.org/abs/2212.08703. Temperature should be applied before softmax, so it cannot be directly implemented inside confidence computation.
- Add a new `ConfidenceEnsembleModel` class that supports running `transcribe` method for the ensemble. It does not support training or any other kind of inference (`forward` method is not implemented). This is by design as we only combine the models through model selection block and it operates on the full transcriptions, not intermediate results.
- Add `scripts/confidence_ensembles/test_confidence_ensembles.py` which contains a few integration tests of building->running ensemble of es-it models. The plan is to include it in the nightly tests, since it depends on ~200mb of data and takes 5-10 minutes to run.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/models/confidence_ensemble.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            2. Compute confidence for each model
+            3. Use logistic regression to pick the "most confident" model
+            4. Return the output of that model
+        """
+        # TODO: lots of duplicate code with building ensemble script
+        aggr_func = get_confidence_aggregation_bank()[self.confidence.aggregation]
+        confidences = []
+        all_transcriptions = []
+        # always requiring to return hypothesis
+        # TODO: make sure to return text only if was False originally
```
[the remark this answers is not in the document]
Comment (Kipok) [reply nemo/collections/asr/models/confidence_ensemble.py:167]: I will resolve this in a follow-up PR(s)

File: nemo/collections/asr/models/rnnt_models.py [hunk 2]
```diff
@@ -442,6 +442,8 @@ def change_decoding_strategy(self, decoding_cfg: DictConfig):
             self.joint.set_loss(self.loss)
```
[the remark this answers is not in the document]
Comment (Kipok) [reply nemo/collections/asr/models/rnnt_models.py:442]: I think those have asr models as direct subblocks, so all changes should already be included. Let me know if that's not the case

File: scripts/confidence_ensembles/build_ensemble.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    # default confidence, can override
+    confidence: ConfidenceConfig = ConfidenceConfig(
+        # we keep frame confidences and apply aggregation manually to get full-utterance confidence
+        preserve_frame_confidence=True,
+        exclude_blank=True,
+        aggregation="mean",
+        method_cfg=ConfidenceMethodConfig(
+            name="entropy",
+            entropy_type="renui",
+            temperature=0.25,  # this is not really temperature, but alpha, see https://arxiv.org/abs/2212.08703
```
[the remark this answers is not in the document]
Comment (Kipok) [reply scripts/confidence_ensembles/build_ensemble.py:95]: That was the plan originally, but it turned out that I cannot support temperature through this config. So temperature is now a parameter of the decoder and if I rename this temperature in here to "alpha", it's going to break existing configs. I can still do that if breaking-change like that is ok. Or I can keep temperature with deprecation warning and add a new parameter for alpha. Or I can just keep it as is for now (although it is becoming confusing). Which way do you prefer?

File: scripts/confidence_ensembles/build_ensemble.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    )
+    temperature: float = 1.0  # this is a real temperature that will be applied to logits
+
+    # this is optional, but can be used to change any aspect of the transcription
+    # config, such as batch size or amp usage. Note that model, data and confidence
+    # will be overriden by this script
+    transcription: transcribe_speech.TranscriptionConfig = transcribe_speech.TranscriptionConfig()
+
+
+def calculate_score(features, labels, pipe):
```
[the remark this answers is not in the document]
Comment (Kipok) [reply scripts/confidence_ensembles/build_ensemble.py:107]: This is a sklearn Pipeline class (as returned from the `train_model_selection` function). I will add all types and docs in a follow-up PR.

File: scripts/confidence_ensembles/build_ensemble.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+def train_model_selection(
+    training_features,
+    training_labels,
+    multi_class="multinomial",
+    C=10000.0,  # disabling regularization by default as overfitting is likely not an issue
+    class_weight="balanced",  # in case training data is imbalanced
+    max_iter=1000,
+):
+    pipe = make_pipeline(
```
[the remark this answers is not in the document]
Comment (Kipok) [reply scripts/confidence_ensembles/build_ensemble.py:126]: I am changing it in a new code version, so let's address that in a new PR

File: nemo/collections/asr/models/rnnt_models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (titu1994) [reply nemo/collections/asr/models/rnnt_models.py:442]: Sounds right, cool

**Review by titu1994 — APPROVED** [review]
Looks good to me

File: nemo/collections/asr/models/confidence_ensemble.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        # TODO: lots of duplicate code with building ensemble script
+        aggr_func = get_confidence_aggregation_bank()[self.confidence_cfg.aggregation]
+        confidences = []
+        all_transcriptions = []
+        # always requiring to return hypothesis
+        # TODO: make sure to return text only if was False originally
+        return_hypotheses = True
+        for model_idx in range(self.num_models):
+            model = getattr(self, f"model{model_idx}")
+            transcriptions = model.transcribe(
```
Comment (fayejf) [inline nemo/collections/asr/models/confidence_ensemble.py:169]: will this support transcribe_partial_audio just like `transcribe_speech.py`?

**Review by fayejf — CHANGES_REQUESTED** [review]
Had a look at` transcribe_speech.py` related part. Overall LGTM. Have two questions.

File: nemo/collections/asr/models/confidence_ensemble.py (hunk 6 above)
Comment (Kipok) [reply nemo/collections/asr/models/confidence_ensemble.py:169]: Not at the moment, right now the assumption is that we transcribe the full thing and then get the confidence. But in principal, there is a plan to add support (likely with a subclass) that will transcribe only part of the utterance and then continue with the transcription of only the single most-confident model

File: examples/asr/transcribe_speech.py [hunk 7]
```diff
@@ -349,6 +354,9 @@ def autocast():
     if type(transcriptions) == tuple and len(transcriptions) == 2:
         transcriptions = transcriptions[0]
 
+    if cfg.return_transcriptions:
+        return transcriptions
```
[the remark this answers is not in the document]
Comment (Kipok) [reply examples/asr/transcribe_speech.py:358]: The confidence ensemble itself is run with return_transcription=False, but inner models will have it set to True. You can have a look at the tests to see how it's used here: https://github.com/NVIDIA/NeMo/blob/83af9d84329441c27622477ab946af8c0b9aa8a2/scripts/confidence_ensembles/test_confidence_ensembles.py#L99

File: nemo/collections/asr/models/confidence_ensemble.py (hunk 6 above)
Comment (fayejf) [reply nemo/collections/asr/models/confidence_ensemble.py:169]: Gotcha. Yeah theoretically all the ASR model (with extended feature) should support transcribing whole utterance and part of it. Please make sure to add this later. Looking forward to it!

File: examples/asr/transcribe_speech.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fayejf) [reply examples/asr/transcribe_speech.py:358]: I would expect this will somehow be integrated with `transcribe_speech.py ` since this is the go-to place for all ASR models offline inference. but this makes sense.

**Review by fayejf — APPROVED** [review]
I've had a look at _transcribe_ part and it looks good to me. Please make sure to add transcribe_partial_audio in the future. Thanks
